File: AppCoordenadasGPS/app/src/main/java/curso/android/appcoordenadasgps/MainActivity.java
package curso.android.appcoordenadasgps;

import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;

import android.Manifest;
import android.annotation.SuppressLint;
import android.content.Context;
import android.content.pm.PackageManager;
import android.location.Location;
import android.location.LocationManager;
import android.os.Bundle;
import android.widget.TextView;
import android.widget.Toast;

import java.text.DecimalFormat;
import java.util.ArrayList;
import java.util.List;

public class MainActivity extends AppCompatActivity {

    String[] requiredPermissions = {Manifest.permission.ACCESS_COARSE_LOCATION, Manifest.permission.ACCESS_FINE_LOCATION};

    public static final int APP_PERMISSIONS_ID = 2023;

    TextView txtValorLatitude, txtValorLongitude;

    double latitude, longitude;
    LocationManager locationManager;
    boolean gpsAtivo = true;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        txtValorLatitude = findViewById(R.id.txtValorLatitude);
        txtValorLongitude = findViewById(R.id.txtValorLongitude);

        locationManager = (LocationManager) getApplication().getSystemService(Context.LOCATION_SERVICE);
        gpsAtivo = locationManager.isProviderEnabled(LocationManager.GPS_PROVIDER);

        if(gpsAtivo) {
            obterCoordenadas();
        } else {
            latitude = 0.00;
            longitude = 0.00;
            setValues();
            showToast("Coordenadas não disponíveis.");
        }
    }

    private void obterCoordenadas() {
        boolean permissaoAtiva = solicitarPermissaoParaObterLocalizacao();
        if(permissaoAtiva){
            capturarUltimaLocalizacaoValida();
        }
    }

    private boolean solicitarPermissaoParaObterLocalizacao() {
        List<String> deniedPermissions = new ArrayList<String>();
        for(String permission : requiredPermissions) {
            if(ContextCompat.checkSelfPermission(MainActivity.this, permission) != PackageManager.PERMISSION_GRANTED){
                deniedPermissions.add(permission);
            }
        }

        if(!deniedPermissions.isEmpty()) {
            ActivityCompat.requestPermissions(MainActivity.this,
                    deniedPermissions.toArray(new String[deniedPermissions.size()]),
                    APP_PERMISSIONS_ID);
        } else {
            return true;
        }
        return false;
    }

    private void capturarUltimaLocalizacaoValida() {
        @SuppressLint("MissingPermission")
        Location location = locationManager.getLastKnownLocation(LocationManager.NETWORK_PROVIDER);

        if(location != null) {
            latitude = location.getLatitude();
            longitude = location.getLongitude();
        } else {
            latitude = 0.00;
            longitude = 0.00;
        }

        setValues();
        showToast("Coordenadas obtidas com sucesso.");
    }

    private void setValues() {
        txtValorLongitude.setText(formatarGeopoint(longitude));
        txtValorLatitude.setText(formatarGeopoint(latitude));
    }

    private void showToast(String message) {
        Toast.makeText(this, message, Toast.LENGTH_SHORT).show();
    }

    private String formatarGeopoint(double value) {
        DecimalFormat decimalFormat = new DecimalFormat("#.#####");
        return decimalFormat.format(value);
    }
}

File: AppCoordenadasGpsKotlin/app/src/main/java/curso/android/appcoordenadasgpskotlin/MainActivity.kt
package curso.android.appcoordenadasgpskotlin

import android.Manifest
import android.annotation.SuppressLint
import android.content.pm.PackageManager
import android.location.LocationManager
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.widget.TextView
import android.widget.Toast
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.core.content.getSystemService
import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.OnMapReadyCallback
import com.google.android.gms.maps.SupportMapFragment
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.MarkerOptions
import java.text.DecimalFormat

class MainActivity : AppCompatActivity(), OnMapReadyCallback {

    private lateinit var mMap: GoogleMap

    var permissoesRequeridas = arrayOf(
        Manifest.permission.ACCESS_FINE_LOCATION,
        Manifest.permission.ACCESS_COARSE_LOCATION
    )

    val APP_PERMISSOES_ID = 2021

    var txtValorLatitude: TextView? = null
    var txtValorLongitude: TextView? = null

    var latitude: Double = 0.00
    var longitude: Double = 0.00

    var locationManager: LocationManager? = null

    var gpsAtivo = true

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        // Obtain the SupportMapFragment and get notified when the map is ready to be used.
        val mapFragment = supportFragmentManager
            .findFragmentById(R.id.map) as SupportMapFragment
        mapFragment.getMapAsync(this)

        txtValorLongitude = findViewById(R.id.txtValorLongitude)
        txtValorLatitude = findViewById(R.id.txtValorLatitude)

        locationManager = application.getSystemService(LOCATION_SERVICE) as LocationManager

        gpsAtivo = locationManager!!.isProviderEnabled(LocationManager.GPS_PROVIDER)

        if(gpsAtivo){
            obterCoordenadas()
        }else{
            latitude = 0.00
            longitude = 0.00

            txtValorLatitude!!.setText(formatarGeopoint(latitude))
            txtValorLongitude!!.setText(formatarGeopoint(longitude))

            Toast.makeText(this, "Coordenadas não disponíveis", Toast.LENGTH_LONG)
        }
    }

    override fun onMapReady(googleMap: GoogleMap) {
        mMap = googleMap

        // Add a marker in Sydney and move the camera
        val sydney = LatLng(latitude, longitude)
        mMap.addMarker(MarkerOptions().position(sydney).title("Celular localizado aqui"))
        mMap.moveCamera(CameraUpdateFactory.newLatLng(sydney))
        mMap.uiSettings.setZoomControlsEnabled(true)
        mMap.setMinZoomPreference(9.5f)
        mMap.setMaxZoomPreference(12.5f)
    }

    private fun obterCoordenadas(){
        val permissaoAtiva: Boolean = solicitarPermissaoObterLocalizacao()
        if(permissaoAtiva)
            capturarUltimaLocalizacaoAtiva()
    }

    private fun solicitarPermissaoObterLocalizacao(): Boolean{
        Toast.makeText(this,"Verificando permissões...", Toast.LENGTH_LONG).show()
        val permissoesNegadas: MutableList<String> = ArrayList()
        var permissaoNegada: Int
        for(permissao in permissoesRequeridas){
            permissaoNegada = ContextCompat.checkSelfPermission(this@MainActivity, permissao)
            if(permissaoNegada != PackageManager.PERMISSION_GRANTED)
                permissoesNegadas.add(permissao)
        }
        return if(permissoesNegadas.isNotEmpty()){
            ActivityCompat.requestPermissions(this@MainActivity, permissoesNegadas.toTypedArray(), APP_PERMISSOES_ID)
            false
        } else {
            true
        }

        return true
    }

    @SuppressLint("MissingPermission")
    private fun capturarUltimaLocalizacaoAtiva(){
        val location = locationManager!!.getLastKnownLocation(LocationManager.GPS_PROVIDER)
        if(location != null) {
            latitude = location.latitude
            longitude = location.longitude
        } else {
            latitude = 0.00
            longitude = 0.00
        }

        txtValorLatitude!!.setText(formatarGeopoint(latitude))
        txtValorLongitude!!.setText(formatarGeopoint(longitude))

        Toast.makeText(this, "Coordenadas obtidas com sucesso", Toast.LENGTH_LONG).show()
    }

    private fun formatarGeopoint(valor: Double): String?{
        val result = DecimalFormat("#.####")
        return result.format(valor)
    }
}